Refined ServiceManager factory, delegator and configuration types, to allow for easier introspection of mis-configuration at type level

boesing opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | kinda
| QA            | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

This reduces the overall baseline by 140 errors. Some errors still have to be suppressed inline.

boesing commented on src/AbstractPluginManager.php:
```diff
@@ -31,6 +33,15 @@
  * The implementation extends `ServiceManager`, thus providing the same set
  * of capabilities as found in that implementation.
+ *
+ * @see ServiceLocatorInterface
```
I use `@see` because the codesniffers out there as well as phpstorm are not able to detect the usage from within `@psalm-import-type`.

boesing commented on src/Config.php:
```diff
@@ -86,32 +89,14 @@ public function toArray()
-        }
-        return $a;
+        /** @psalm-suppress MixedReturnTypeCoercion */
+        return ArrayUtils::merge($a, $b);
```
I removed the inlined code here as `laminas-stdlib` is a dependency within `ServiceManager` to use `merge` for the `lazy_services`.

weierophinney replied:
:+1:

boesing commented on src/ServiceManager.php:
```diff
@@ -271,40 +291,8 @@ public function getAllowOverride()
-     *
-     * Valid top keys are:
-     *
-     * - services: service name => service instance pairs
```
these details are obsolete as we are now providing a proper psalm type.
so imho this can be deleted, right?

weierophinney replied:
I would say "yes', because:

- We have end-user documentation detailing what is accepted for each.
- The Psalm types give IDEs the information necessary to provide hinting to developers.

boesing commented on src/ServiceManager.php:
```diff
@@ -415,8 +404,11 @@ public function setInvokableClass($name, $class = null)
-     *     to map.
+     * @param string|callable|Factory\FactoryInterface $factory  Factory to which to map.
+     * phpcs:disable Generic.Files.LineLength.TooLong
+     * @psalm-param class-string<Factory\FactoryInterface>|callable(ContainerInterface,string,ServiceOptionsType|null)|Factory\FactoryInterface $factory
```
it seems that `@psalm-param` cannot properly detect multiline declaration and thus, I had to use `phpcs:disable` to disable the line length check. 
Screw you, line length check!

At least I was not able to do it here, other methods worked obviously. Not sure what the reason is but TBH, I don't really care and I don't want to spend more time on this just to make the codestyle happy.

boesing commented on src/AbstractPluginManager.php:
```diff
@@ -31,6 +33,15 @@
+ *
+ * @see ServiceLocatorInterface
+ *
+ * @template TPlugin of object
```
So finally we can start providing generics for plugin managers

boesing commented on src/PluginManagerInterface.php:
```diff
@@ -11,6 +11,9 @@
  * A plugin manager is a specialized service locator used to create homogeneous objects
+ *
+ * @template TPlugin of object
+ * @method build($name, ?array $options = null): TPlugin
```
TBH, I did not verify if that really works.

boesing commented on src/ServiceManager.php:
```diff
@@ -78,29 +86,39 @@ class ServiceManager implements ServiceLocatorInterface
      * A list of already loaded services (this act as a local cache)
      *
-     * @var array
+     * @var array<string,mixed>
```
I've seen somewhere that only `array|object` is allowed for services. Maybe something to consider here?

boesing replied:
https://github.com/laminas/laminas-servicemanager/pull/98/files#diff-c88e8922673ece50d04aef21f4dc95b87bd465868eb073480467f87e626e3b0aR478

weierophinney replied:
Yeah - only array or object are allowed as return types.

boesing replied:
~but then we have to bc break `ServiceManager::doCreate` as it allows to return `mixed`~

its private. I'll add native type-hint which might break stuff upstream as it is definitely possible to return any type from a factory even tho it states one has to return objects.

boesing commented on src/ConfigInterface.php:
```diff
@@ -4,6 +4,56 @@
+ *     initializers?: InitializersConfigurationType,
+ *     invokables?: array<string,string>,
+ *     lazy_services?: LazyServicesConfigurationType,
+ *     services?:array<string,object>&array{config?:ArrayAccess<string,mixed>|array<string,mixed>},
```
https://github.com/laminas/laminas-servicemanager/pull/98/files#diff-c88e8922673ece50d04aef21f4dc95b87bd465868eb073480467f87e626e3b0aR478

This allows `array|object`.
Yes, I know that this would not only allow `config` but also other arrays.
As @Ocramius pointed out in #99, this might be a BC break (even tho, we still accept arrays), so should I change this to `array<string,object|array>` instead?

weierophinney replied:
yes.

weierophinney replied:
yes.

weierophinney commented on src/AbstractPluginManager.php:
```diff
@@ -108,14 +118,13 @@ public function __construct($configInstanceOrParentLocator = null, array $config
     }
 
     /**
-     * Override configure() to validate service instances.
```
This should still be in the docblock, as it gives a description of what we are doing here while overriding the parent.

weierophinney commented on src/ServiceManager.php:
```diff
@@ -412,8 +403,11 @@ public function setInvokableClass($name, $class = null)
-     * @param string|callable|Factory\FactoryInterface $factory Factory to which
-     *     to map.
+     * @param string|callable|Factory\FactoryInterface $factory  Factory to which to map.
+     * phpcs:disable Generic.Files.LineLength.TooLong
+     * @psalm-param class-string<Factory\FactoryInterface>|callable(ContainerInterface,string,array<mixed>|null)|Factory\FactoryInterface $factory
+     * phpcs:enable Generic.Files.LineLength.TooLong
```
```suggestion
     * @psalm-param class-string<Factory\FactoryInterface>
     *     |callable(ContainerInterface,string,array<mixed>|null)
     *     |Factory\FactoryInterface $factory
```

( Matches what you did on lines 453-454, and eliminates the need for the phpcs directives)

boesing replied:
When I do this locally, I get a psalm error.

```
ERROR: PropertyTypeCoercion - src/ServiceManager.php:421:9 - $this->factories expects 'array<string, Laminas\ServiceManager\Factory\FactoryInterface|callable(Interop\Container\ContainerInterface, string, array<array-key, mixed>)|class-string>',  parent type 'non-empty-array<string, Laminas\ServiceManager\Factory\FactoryInterface|callable|callable(Interop\Container\ContainerInterface, string, array<array-key, mixed>)|string>' provided (see https://psalm.dev/198)
        $this->factories[$name] = $factory;
```

Somehow there is an additional `callable` in there. I guess its due to the `null` type-hint

weierophinney commented on src/ServiceManager.php:
```diff
@@ -578,55 +581,38 @@ private function createDelegatorFromName($name, ?array $options = null)
-                    DelegatorFactoryInterface::class
-                ));
-            }
+            $this->assertCallableDelegatorFactory($delegatorFactory);
```
:+1:

Love the method extraction.

weierophinney commented on src/ServiceManager.php:
```diff
@@ -968,4 +958,33 @@ private function abstractFactoryCanCreate(string $name): bool
+     */
+    private function assertCallableDelegatorFactory($delegatorFactory): void
+    {
+        if ($delegatorFactory instanceof Factory\DelegatorFactoryInterface) {
+            return;
+        }
+
+        if (! is_callable($delegatorFactory)) {
+            if (is_string($delegatorFactory)) {
+                throw new ServiceNotCreatedException(sprintf(
+                    'An invalid delegator factory was registered; resolved to class or function "%s"'
+                    . ' which does not exist; please provide a valid function name or class name resolving'
+                    . ' to an implementation of %s',
+                    $delegatorFactory,
+                    DelegatorFactoryInterface::class
+                ));
+            }
+
+            throw new ServiceNotCreatedException(sprintf(
+                'A non-callable delegator, "%s", was provided; expected a callable or instance of "%s"',
+                is_object($delegatorFactory) ? get_class($delegatorFactory) : gettype($delegatorFactory),
+                DelegatorFactoryInterface::class
+            ));
+        }
+    }
```
```suggestion
        if ($delegatorFactory instanceof Factory\DelegatorFactoryInterface
            || is_callable($delegatorFactory)
        ) {
            return;
        }

        if (is_string($delegatorFactory)) {
            throw new ServiceNotCreatedException(sprintf(
                'An invalid delegator factory was registered; resolved to class or function "%s"'
                . ' which does not exist; please provide a valid function name or class name resolving'
                . ' to an implementation of %s',
                $delegatorFactory,
                DelegatorFactoryInterface::class
            ));
        }

      throw new ServiceNotCreatedException(sprintf(
            'A non-callable delegator, "%s", was provided; expected a callable or instance of "%s"',
            is_object($delegatorFactory) ? get_class($delegatorFactory) : gettype($delegatorFactory),
            DelegatorFactoryInterface::class
        ));
    }
```

Basically, return early as soon as we know it's valid, and otherwise just vary the exception based on whether or not it was passed as a string.

weierophinney commented on test/AbstractPluginManagerTest.php:
```diff
@@ -365,7 +365,7 @@ public function testValidateWillFallBackToValidatePluginWhenDefinedAndEmitDeprec
 
         $errorHandlerCalled = false;
-        set_error_handler(function ($errno, $errmsg) use (&$errorHandlerCalled) {
+        set_error_handler(function (int $errno, string $errmsg) use (&$errorHandlerCalled) {
```
```suggestion
        set_error_handler(function (int $errno, string $errmsg) use (&$errorHandlerCalled): void {
```

Add the return type.

weierophinney commented on test/CommonServiceLocatorBehaviorsTrait.php:
```diff
@@ -442,8 +448,9 @@ public function testCanConfigureAllServiceTypes(): void
                 stdClass::class => [
-                    function ($container, $name, $callback) {
-                        $instance      = $callback();
+                    function (ContainerInterface $container, string $name, callable $callback) {
```
```suggestion
                    function (ContainerInterface $container, string $name, callable $callback): object {
```

weierophinney commented on test/CommonServiceLocatorBehaviorsTrait.php:
```diff
@@ -899,7 +910,8 @@ public function testAllowOverrideFlagIsMutable(ServiceManager $container): void
         $container = $this->createContainer();
-        set_error_handler(function ($errno, $errstr) {
+        /** @psalm-suppress InvalidArgument */
+        set_error_handler(function (int $errno) {
```
```suggestion
        set_error_handler(function (int $errno): void {
```

weierophinney reviewed (dismissed):
I think this is a GREAT addition! I've flagged a few nitpicks, which are primarily to close a few more Psalm errors as well as simplify some of the methods by promoting early returns.

Otherwise, :ship:

boesing committed:
bugfix: widen `services` type as requested in code review

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: restore method description

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: limit services to be either `array|object` as outlined in code review

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: add `@psalm-require-extends` to ensure trait is only used within a unit test

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactor: optimize callable delegator assertion

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add return types to some functions

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
After some talks to @Ocramius, we are thinking about extracting the `@template`-Annotation into a dedicated patch to avoid upstream analysis errors.

The dedicated patch would target v4.0 which does not have any ETA and which would require a huge amount of work in a majority of our components. Maybe this is something for the hacktoberfest?

Ocramius commented on src/AbstractPluginManager.php:
```diff
@@ -31,6 +33,11 @@
  * The implementation extends `ServiceManager`, thus providing the same set
  * of capabilities as found in that implementation.
+ *
+ * @template TPlugin of object
```
As discussed in chat, probably worth dropping this from this specific patch

boesing replied:
Removed from this patch.

Ocramius commented on src/Config.php:
```diff
@@ -86,32 +89,14 @@ public function toArray()
+     * @psalm-param ServiceManagerConfigurationType $a
+     * @psalm-param ServiceManagerConfigurationType $b
+     * @psalm-return ServiceManagerConfigurationType
+     * @psalm-suppress MixedReturnTypeCoercion
```
This suppression should only be in either the body of the method, or the docblock, but not both

boesing replied:
it actually has to be there. if you check the psalm config, I've added "report unused psalm suppress".

Ocramius commented on src/ConfigInterface.php:
```diff
@@ -4,6 +4,56 @@
+ *      array<
+ *          array-key,
+ *          (class-string<Factory\DelegatorFactoryInterface>|Factory\DelegatorFactoryInterface)
+ *          |class-string
```
Why is any `class-string` allowed?

boesing replied:
Good catch. Its not.

Ocramius commented on src/ConfigInterface.php:
```diff
@@ -4,6 +4,56 @@
+ * @psalm-type FactoriesConfigurationType = array<
+ *      string,
+ *      (class-string<Factory\FactoryInterface>|Factory\FactoryInterface)
+ *      |class-string
```
Why is any `class-string` allowed?

Ocramius commented on src/Factory/DelegatorFactoryInterface.php:
```diff
@@ -23,6 +23,7 @@ interface DelegatorFactoryInterface
      * @param  string             $name
+     * @psalm-param callable():mixed $callback
      * @param  null|array         $options
      * @return object
```
If this is supposed to return `object`, then `$callback` cannot return `mixed`.

I feel like the `DelegatorFactoryInterface` may benefit from a templated type, in future (not here, due to BC)

boesing replied:
Since `ContainerInterface#get` can return `mixed`, this actually has to be `mixed` :-(

But yes, having a templated type would help for sure but until then, I've thought that `mixed` would be fine until the return type is object 🤷🏼‍♂️

Ocramius replied:
Yeah, but the commented line is about `@return object` (wrong)

boesing replied:
Not sure if I can follow.

Do you want me to change existing documentation that we end up having bc breaks? as of now, due to the lack of typing, callable factories can return mixed. When I enforce `object` now, this will be pain probably. Thats why I think that might be a BC break but I am not 100% sure tho.

Or do you want me to open up the return value of the delegator to be mixed?

boesing replied:
I think I am fine with both as I don't have that much experience with adding psalm types on such a widely used component. As always, I trust your experience. :-)

Ocramius commented on src/PluginManagerInterface.php:
```diff
@@ -11,6 +11,9 @@
  * A plugin manager is a specialized service locator used to create homogeneous objects
+ *
+ * @template TPlugin of object
+ * @method build($name, ?array $options = null): TPlugin
```
This method really shouldn't be declared here, since otherwise we require child implementations to include a `build()` method, or face scenarios where the child class does not implement it, yet psalm considers it existing => runtime crash.

Let's drop it.

boesing replied:
The method is declared in the `ServiceLocatorInterface` which is extended from this interface.
I just wanted to tighten the return type of the `build` method but I forgot to add `get`-method as well.

Will do so in the patch for the next major. until then, I drop it.

Ocramius commented on src/ConfigInterface.php:
```diff
@@ -4,6 +4,56 @@
+ *      string,
+ *      (class-string<Factory\FactoryInterface>|Factory\FactoryInterface)
+ *      |class-string
+ *      |callable(ContainerInterface,string,array<mixed>)
```
I think `array<mixed>|null` is needed, to allow for factories to deal with the "no config provided" scenario?

Ocramius commented on src/ConfigInterface.php:
```diff
@@ -4,6 +4,56 @@
+ *          array-key,
+ *          (class-string<Factory\DelegatorFactoryInterface>|Factory\DelegatorFactoryInterface)
+ *          |class-string
+ *          |callable(ContainerInterface,string,callable():object,array<mixed>):object
```
I think `array<mixed>|null` is needed, to allow for delegators to deal with the "no config provided" scenario?

boesing replied:
good catch

Ocramius commented on src/ServiceManager.php:
```diff
@@ -455,6 +450,8 @@ public function addAbstractFactory($factory)
      * @param string|callable|Factory\DelegatorFactoryInterface $factory Delegator
      *     factory to assign.
+     * @psalm-param class-string<Factory\DelegatorFactoryInterface>
+     *     |callable(ContainerInterface,string,callable,array<mixed>) $factory
```
I think `array<mixed>|null` is needed

Ocramius commented on src/ServiceManager.php:
```diff
@@ -412,8 +403,11 @@ public function setInvokableClass($name, $class = null)
-     *     to map.
+     * @param string|callable|Factory\FactoryInterface $factory  Factory to which to map.
+     * phpcs:disable Generic.Files.LineLength.TooLong
+     * @psalm-param class-string<Factory\FactoryInterface>|callable(ContainerInterface,string,array<mixed>|null)|Factory\FactoryInterface $factory
```
Should we declare some of the more complex union types such as `class-string<Factory\FactoryInterface>|callable(ContainerInterface,string,array<mixed>|null)|Factory\FactoryInterface` as `FactoryType` in psalm, or such?

boesing replied:
TBH; I am happy that psalm is green. I've now spent around 1 hour to extract some types into psalm types without success.

Would prefer doing so in another patch.

Ocramius replied:
Yeah, it is an isomorphic change anyway 👍

Ocramius commented on src/ServiceManager.php:
```diff
@@ -507,16 +507,20 @@ public function setShared($name, $flag)
-     * @param string[]|Initializer\InitializerInterface[]|callable[] $initializers
+     * @psalm-param array<
+     *     array-key,
+     *     class-string<Initializer\InitializerInterface>|Initializer\InitializerInterface|callable
```
The `callable` type here should be typed better: removes the need for declaring  `@psalm-var` below

Ocramius commented on src/ServiceManager.php:
```diff
@@ -507,16 +507,20 @@ public function setShared($name, $flag)
             }
 
             if (is_callable($initializer)) {
+                /** @psalm-var callable(ContainerInterface,object):void */
```
Becomes unnecessary if we type the above  `callable` a bit better

Ocramius commented on src/ServiceManager.php:
```diff
@@ -543,9 +546,11 @@ private function getFactory($name)
         }
 
         if (is_callable($factory)) {
+            /** @psalm-var callable(ContainerInterface,string,array<mixed>|null):object $factory */
```
shouldn't be needed? Perhaps this is because of the missing `|null` in above factory definitions (for the options parameter)?

boesing replied:
Nah, the problem is, that a few lines above, `$factory` is instantiated due to `class_exists` and thus, it *probably* could be any object implementing `__invoke` (which is actually true).

And thus, psalm cannot properly infer that the instantiated object is `__callable` while following the rules of the `callable` annotation.

---

Took me a bit until I understood this: https://github.com/vimeo/psalm/issues/5384

Ocramius commented on src/ServiceManager.php:
```diff
@@ -578,55 +581,37 @@ private function createDelegatorFromName($name, ?array $options = null)
+            $this->assertCallableDelegatorFactory($delegatorFactory);
 
+            /**
+             * @psalm-var callable(ContainerInterface,string,callable,array<mixed>|null):object $delegatorFactory
```
Should not be needed, especially after  `psalm-assert`

Ocramius commented on src/ServiceManager.php:
```diff
@@ -578,55 +581,37 @@ private function createDelegatorFromName($name, ?array $options = null)
+             */
             $this->delegators[$name][$index] = $delegatorFactory;
 
             $creationCallback = function () use ($delegatorFactory, $name, $creationCallback, $options) {
```
I think this closure should be `static` - we're probably leaking GC (and memory, obviously) here...

Ocramius commented on src/ServiceManager.php:
```diff
@@ -578,55 +581,37 @@ private function createDelegatorFromName($name, ?array $options = null)
      * @throws ContainerException If any other error occurs.
      */
-    private function doCreate($resolvedName, ?array $options = null)
+    private function doCreate(string $resolvedName, ?array $options = null): object
```
This will break with `array` services, due to the newly declared `: object` return value

boesing replied:
`array` services are not retrievable via `doCreate` as this method only uses `factories` or `abstract_factories` which both must return `object` as per definition.

`array` is "usually" only `ServiceManager#setService('config', $configArray);` or somehow else but it actually is not possible to create an array via a factory.

anyways, I've converted this to a soft type-hint

Ocramius commented on test/ConfigTest.php:
```diff
@@ -106,6 +106,7 @@ public function testPassesKnownServiceConfigKeysToServiceManagerWithConfigMethod
         $services = $this->prophesize(ServiceManager::class);
         $services->configure($expected)->willReturn('CALLED');
 
+        /** @psalm-suppress InvalidArgument */
```
Let's add a comment here - ideally, test should be adapted, but it's fine to show that the system keeps working as it did before

boesing replied:
I've added a comment. I already forgot why I've added this initially TBH.

Ocramius commented on test/ContainerTest.php:
```diff
@@ -6,15 +6,22 @@
 
     protected function createContainer(array $config): ContainerInterface
     {
+        /** @psalm-suppress MixedArgumentTypeCoercion */
```
The base `createContainer` method should use the `ServiceManagerConfigurationType`, but I see it's imported here, and never used

boesing replied:
I can either suppress this error or `MoreSpecificImplementedParamType` due to the implementation details here: 
https://github.com/laminas/laminas-container-config-test/blob/6a7c3f3fed29ed844cdc22b1d911cb2ef745873e/src/AbstractContainerTest.php#L18

I don't know what would be better tho.

Ocramius replied:
This was more a note on the fact that the imported type in this file is not used. The class-level docblock can probably be dropped

boesing replied:
Ah yes, I've added it initially but then removed it again. Will drop that.

Ocramius commented on test/ServiceManagerTest.php:
```diff
@@ -475,7 +487,6 @@ public function testConfigureMultipleTimesAvoidsDuplicates(): void
         ];
 
         $serviceManager = new ServiceManager($config);
-        $serviceManager->configure($config);
```
This test is explicitly called `testConfigureMultipleTimesAvoidsDuplicates` - it's designed to do the configuration multiple times, so this call should stay

boesing committed:
feature: introduce plenty of psalm types

This reduces the overall baseline by 140 errors. Some errors still have
to be suppressed inline.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: remove `class-string` type from some config entries

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: allow `null` for the `$options` argument in `delegators` and `factories` config entries

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: allow `null` for `$options` in delegator callables

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: re-use psalm type rather than re-declaring it

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove configuration type for the deprecated `DelegatorFactoryInterface`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove unnecessary var annotation

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactor: make nested creation callback static

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove native return type from `ServiceManager#doCreate`

This will prevent introducing BC breaks while still provide a soft return type.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qb: add detailed reason on why we suppress the invalid argument

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: restore `ServiceManager#configure` call

This will ensure the test does exactly what it supposed to do.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented on src/AbstractPluginManager.php:
```diff
@@ -110,16 +116,18 @@ public function __construct($configInstanceOrParentLocator = null, array $config
     public function configure(array $config)
     {
         if (isset($config['services'])) {
+            /** @psalm-suppress MixedAssignment */
```
This should be fixable: `@psalm-param ServiceManagerConfiguration|array{services: ServiceManagerConfiguration} $config`

boesing replied:
That would be a false type as `services` does not contain another `ServiceManagerConfiguration`. 
It contains `array<string,object|array>` which is properly annotated but somehow psalm refuses to infer it from the parameter type.

```php
/**
 * @psalm-type ServiceManagerConfigurationType = array{
 *     services?: array<string,object|array>
 * }
 */
```

Ocramius replied:
Ah, I misread it - thought it was a different nesting level

Ocramius commented on src/AbstractPluginManager.php:
```diff
@@ -167,7 +173,9 @@ public function get($name, ?array $options = null)
         }
 
-        $instance = empty($options) ? parent::get($name) : $this->build($name, $options);
+        $options = $options ?? [];
```
Instead of creating a new array (expensive), let's check if it is `null`, and call `build()` if it is **not** null? I would also expect an explicit call to `build()` even if an empty array is given...

Ocramius replied:
Keeping `empty()` could also be a sufficient solution for now

boesing replied:
I've added `!$options` - that would avoid calling `empty` and does not require a new array.
Would that be sufficient?

Ocramius commented on src/Config.php:
```diff
@@ -26,10 +22,12 @@
 class Config implements ConfigInterface
 {
-    /** @var array */
+    /** @var array<string,bool> */
```
Not important, but I noticed that your types have no spacing in it 😁

boesing replied:
why would they need spacings? is that another coding style thing? 
The more spaces, the merrier line length issues on complex types 😩

Ocramius replied:
Types are not just an afterthought - should be as readable as other stuff 😛

Ocramius commented on src/Config.php:
```diff
@@ -66,6 +67,8 @@ public function __construct(array $config = [])
             }
         }
+
+        /** @psalm-suppress ArgumentTypeCoercion */
```
This coercion shouldn't apply anymore: `merge()` has the correct types? 🤔

boesing replied:
`ArrayUtils::merge` returns `array` and thus, psalm loses track.

Not sure if something like this would help psalm to keep track:

```
/**
 * @template TArray1 of array
 * @template TArray2 of array
 * @param TArray1 $a
 * @param TArray2 $b
 * @return TArray1&TArray2
 */
```

But this wont properly work for nested arrays (as they are here). There is absolutely no way but to trust that `ArrayUtils::merge` will return the same types.

Ocramius replied:
Yes, but you are not calling `ArrayUtils::merge()`, but rather `$this->merge()`, which is well typed, and has its own suppressions

Ocramius replied:
I'll leave this alone for now - I think it's not worth wasting effort on it after all the work you've already put in this, and it's in the internals anyway :)

boesing replied:
I've added `reportUnusedPsalmSuppress` in the config and thats not reported so something might still be odd :-/

Ocramius commented on src/PluginManagerInterface.php:
```diff
@@ -17,11 +17,12 @@ interface PluginManagerInterface extends ServiceLocatorInterface
      * @throws InvalidServiceException If created instance does not respect the
      *     constraint on type imposed by the plugin manager.
      * @throws ContainerException If any other error occurs.
+     * @psalm-assert TPlugin $instance
```
Leftover?

Ocramius commented on src/ServiceManager.php:
```diff
@@ -563,70 +564,58 @@ private function getFactory($name)
-            };
+            $creationCallback =
+                /** @return object */
+                static function () use (
```
Making this a static function is a huge improvement!

boesing committed:
qa: remove unused psalm type import

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactor: avoid creating a new option array in pluginmanager if not needed

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove leftover template type

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented on test/ServiceManagerTest.php:
```diff
@@ -20,16 +21,23 @@
 use function get_class;
 
 /**
+ * @see ConfigInterface
```
I really like how you used `@see` to import types that are otherwise elided by the CS tooling :D

Ocramius commented:
🚢 

will also release ASAP - it's a huge improvement :)
